src/Runtime/Core/Collections/ReadOnlyArray.cs

```diff
@@ -0,0 +1,193 @@
+    }
+
+    /// <summary>
+    /// A version if <see cref="ArraySegment{T}"/> which can not be mutated.
+    /// </summary>
+    public struct ReadOnlyArraySegment<T> : IReadOnlyList<T>
+    {
+        /// <summary>
+        /// Underlying read-only array.
+        /// </summary>
+        private readonly ReadOnlyArray<T> array;
+
+        /// <summary>
+        /// Index of the first element which belongs to this segment.
+        /// </summary>
+        private readonly int begin;
+
+        /// <summary>
+        /// Index of the first element which does not belong to this segment.
+        /// </summary>
+        private readonly int end;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ReadOnlyArraySegment{T}"/> structure.
+        /// </summary>
+        public ReadOnlyArraySegment(ReadOnlyArray<T> array, int begin, int end)
+        {
+            this.array = array;
+            this.begin = begin;
+            this.end = end;
+        }
+
+        /// <inheritdoc/>
+        public T this[int index] => this.array[this.begin + index];
+
+        /// <inheritdoc/>
+        public int Count => this.end - this.begin;
+
+        /// <summary>
+        /// Returns enumerator over elements of array.
+        /// </summary>
+        /// <remarks>
+        /// This is value-type non-virtual version of enumerator that is used by compiler in foreach loops.
+        /// </remarks>
+        public ReadOnlySegmentEnumerator<T> GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this.array, this.begin, this.end);
+
+        /// <inheritdoc/>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this.array, this.begin, this.end);
+
+        /// <inheritdoc/>
+        IEnumerator IEnumerable.GetEnumerator() =>
+            new ReadOnlySegmentEnumerator<T>(this.array, this.begin, this.end);
+    }
+
+    /// <summary>
+    /// Enumerator for read only arrays and read only array segments.
+    /// </summary>
+    public struct ReadOnlySegmentEnumerator<T> : IEnumerator<T>
+    {
+        /// <summary>
+        /// Underlying read-only array.
+        /// </summary>
+        private readonly ReadOnlyArray<T> array;
+
+        /// <summary>
+        /// Index of the first element which belongs segment begin enumerated.
+        /// </summary>
+        private readonly int begin;
+
+        /// <summary>
+        /// Index of the first element which does not belong segment begin enumerated.
+        /// </summary>
+        private readonly int end;
+
+        /// <summary>
+        /// Index of the current element
+        /// </summary>
+        private int pointer;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ReadOnlyArraySegment{T}"/> structure.
+        /// </summary>
+        public ReadOnlySegmentEnumerator(ReadOnlyArray<T> array, int begin, int end)
```

Comment:
Given that this constructor is public, I would suggest some basic error checking be applied.  Some simple null and bounds checking would suffice.

Reply:
Made this constructor internal, didn't add any checks.